File: 2020/Day_06/code.php
<?php

require_once __DIR__ . '/../../lib/php/utils.php';

$input = readInput(__DIR__, PHP_EOL . PHP_EOL);

// Task code
function part01(array $input): int
{
    $sum = 0;

    foreach ($input as $value) {
        $sum += count(
            array_unique(
                str_split(
                    str_replace(PHP_EOL, '', $value)
                )
            )
        );
    }

    return $sum;
}

function part02(array $input): int
{
    $sum = 0;

    foreach ($input as $line) {
        $group = explode(PHP_EOL, $line);
        $groupCount = count($group);
        $questionList = [];

        foreach ($group as $person) {
            foreach (str_split($person) as $question) {
                $questionList[$question] = ($questionList[$question] ?? 0) + 1;
            }
        }

        $sum += count(array_filter($questionList, static fn ($answers) => $answers === $groupCount));
    }

    return $sum;
}

// Execute
calcExecutionTime();
$result01 = part01($input);
$result02 = part02($input);
$executionTime = calcExecutionTime();

writeln('Solution Part 1: ' . $result01);
writeln('Solution Part 2: ' . $result02);
writeln('Execution time: ' . $executionTime);

saveBenchmarkTime($executionTime, __DIR__);

// Task test
testResults(
    [6748, 3445], // Expected
    [$result01, $result02], // Result
);

File: 2021/Day_14/code.go
package main

import (
	"advent-of-code/helper"
	"fmt"
	"math"
	"path"
	"runtime"
	"strings"
	"time"
)

func main() {
	_, f, _, _ := runtime.Caller(0)
	cwd := path.Join(path.Dir(f))

	input := helper.ReadInput(cwd, helper.NewLine)
	iValues := input.Strings()

	// Execute
	start := time.Now()
	result01 := part01(iValues)
	result02 := part02(iValues)
	executionTime := helper.ExecutionTime(time.Since(start))

	fmt.Printf("Solution Part 1: %v\n", result01)
	fmt.Printf("Solution Part 2: %v\n", result02)
	fmt.Printf("Execution time: %s\n", executionTime)

	helper.SaveBenchmarkTime(executionTime, cwd)

	// Testing
	helper.TestResults(
		[]helper.TestingValue{
			helper.TestingValue{Result: result01, Expect: 3831},
			helper.TestingValue{Result: result02, Expect: 5725739914282},
		},
	)
}

// Task code
func part01(input []string) int {
	return calculate(input, 10)
}

func part02(input []string) int {
	return calculate(input, 40)
}

func calculate(input []string, maxSteps int) int {
	template := input[0]
	var pairList []string
	polymerFreq := make(map[string]int, 0)
	rules := readRules(input[2:])

	for i := 0; i < len(template)-1; i++ {
		pair := template[i : i+2]

		if _, ok := polymerFreq[pair]; !ok {
			pairList = append(pairList, pair)
		}

		polymerFreq[pair]++
	}

	for steps := 0; steps < maxSteps; steps++ {
		pairList, polymerFreq = countFrequency(pairList, polymerFreq, rules)
	}

	count := make(map[string]int, 0)
	for p, o := range polymerFreq {
		count[string(p[0])] += o
		count[string(p[1])] += o
	}

	count[string(template[0])]++
	count[string(template[len(template)-1])]++

	for c, o := range count {
		count[c] = int(math.RoundToEven(float64(o / 2)))
	}

	least, most := leastAndMostCharCount(count)

	return most - least
}

func readRules(input []string) map[string]string {
	pm := make(map[string]string, 0)

	for _, l := range input {
		v := strings.Split(l, " -> ")

		pm[v[0]] = v[1]
	}

	return pm
}

func countFrequency(pairList []string, freq map[string]int, rules map[string]string) ([]string, map[string]int) {
	nFreq := copyMap(freq)
	nPairList := pairList

	for _, p := range pairList {
		for from, to := range rules {
			if p == from {
				nPStart := string(p[0]) + to
				nPEnd := to + string(p[1])

				_, startExists := nFreq[nPStart]
				if !startExists {
					nPairList = append(nPairList, nPStart)
				}

				_, endExists := nFreq[nPEnd]
				if !endExists {
					nPairList = append(nPairList, nPEnd)
				}

				o := freq[p]
				nFreq[p] -= o
				nFreq[nPStart] += o
				nFreq[nPEnd] += o

				break
			}

		}
	}

	return nPairList, nFreq
}

func leastAndMostCharCount(charMap map[string]int) (least int, most int) {
	for _, v := range charMap {
		least = v
		break
	}

	for _, amount := range charMap {
		if amount < least {
			least = amount
		}

		if amount > most {
			most = amount
		}
	}

	return least, most
}

func copyMap(h map[string]int) map[string]int {
	c := make(map[string]int, 0)

	for k, v := range h {
		c[k] = v
	}

	return c
}
